definition_validator/validator.cc

```diff
@@ -942,17 +938,68 @@ class ValidateWalk {
             return;
         }
 
+        vector<core::MethodRef> missingMethods;
         for (auto proto : abstract) {
             if (proto.data(gs)->owner == sym) {
                 continue;
             }
 
-            auto mem = sym.data(gs)->findConcreteMethodTransitive(gs, proto.data(gs)->name);
-            if (!mem.exists()) {
-                if (auto e = gs.beginError(loc, core::errors::Resolver::BadAbstractMethod)) {
-                    e.setHeader("Missing definition for abstract method `{}`", proto.show(gs));
-                    e.addErrorLine(proto.data(gs)->loc(), "defined here");
-                    e.addAutocorrect(suggestMethodDefinition(gs, loc, proto));
+            auto concreteMethodRef = sym.data(gs)->findConcreteMethodTransitive(gs, proto.data(gs)->name);
+            if (concreteMethodRef.exists()) {
+                continue;
+            }
+
+            missingMethods.emplace_back(proto);
+        }
+
+        if (!missingMethods.empty()) {
+            if (auto e = gs.beginError(classOrModuleDeclaredAt, core::errors::Resolver::BadAbstractMethod)) {
+                auto pluralization = (missingMethods.size() > 1 ? "s" : "");
+                e.setHeader("Missing definition{} for abstract method{}", pluralization, pluralization);
+
+                auto classOrModuleEndsAt =
+                    core::Loc{classOrModuleDeclaredAt.file(), classDef.loc.copyEndWithZeroLength()};
+                auto hasSingleLineDefinition =
+                    classOrModuleDeclaredAt.position(gs).first.line == classOrModuleEndsAt.position(gs).second.line;
+
+                auto endRange = classOrModuleDeclaredAt.copyEndWithZeroLength().join(classOrModuleEndsAt);
+                string endSuffix(endRange.source(gs).value_or(""));
+                regex emptyBodyRe("^\\s*;\\s+end\\s*$");
+                auto hasEmptyBody = regex_match(endSuffix, emptyBodyRe);
```

Comment:
Is it not enough to skip the regex stuff and instead check `classDef->rhs.empty()`?

Reply:
Ah, it looks like at this point we've created a fake `<static-init>` method in the method body:

```
(lldb) p (void)(printf("%s\n", tree.toString(ctx).c_str()))
class ::Child<<C Child>> < (::Parent)
  def self.<static-init>(<blk>)
    <emptyTree>
  end
end
```

which means that this code would need to be something like

```cpp
auto hasEmptyBody = classDef.rhs.empty();
if (classDef.rhs.size() == 1) {
    if (auto *mdef = ast::cast_tree<ast::MethodDef>(classDef.rhs[0])) {
        hasEmptyBody = ast::isa_tree<ast::EmptyTree>(mdef->rhs);
    }
}
```

which is more code but less reliant on regex.

Reply:
I must admit that I don't yet understand what we expect the AST to look like at various stages, but your suggestion passed all of the tests as written, so I applied it in 45418ae.